### ISSUE TITLE
SetupBuiltinSymbolTable bug when switching between SPIR-V and OpenGL

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/glslang#355]
We use glslang in a distributed shader build system where a worker might switch between compiling SPIR-V and OpenGL shaders.  I have been hunting down a crash with an invalid symbolTable, and I have figured out what the problem is and here I propose a fix (although I think it might need some modification to be more robust).

The basic problem is that SetupBuiltinSymbolTable caches the built-in symbols in two multi-dimensional arrays CommonSymbolTable and SharedSymbolTable.  However, these built-in symbol tables are not paramaterized based on what SpvVersion is set.  This is a problem because the symbol tables *do* depend on the SpvVersion.

My fix, shown below, is to compute a 0 or 1 index based on SpvVersion.  I am not certain whether this is sufficient because it may need to be further paramaterized based on what values are in SpvVersion.

Here is a diff of my fix:

```
==== //source2/main/src/thirdparty/glslang/glslang/MachineIndependent/ShaderLang.cpp#21 - H:/dev/source2/main\src\thirdparty\glslang\glslang\MachineIndependent\ShaderLang.cpp ====
120a121,133
> int MapSpvVersionToIndex( const SpvVersion &spvVersion )
> {
> 	if ( spvVersion.spv != 0 )
> 	{
> 		return 1;
> 	}
> 	else
> 	{
> 		return 0;
> 	}
> }
> const int SpvProfileCount = 2;
> 
136,137c149,150
< TSymbolTable* CommonSymbolTable[VersionCount][ProfileCount][EPcCount] = {};
< TSymbolTable* SharedSymbolTables[VersionCount][ProfileCount][EShLangCount] = {};
---
> TSymbolTable* CommonSymbolTable[VersionCount][ProfileCount][SpvProfileCount][EPcCount] = {};
> TSymbolTable* SharedSymbolTables[VersionCount][ProfileCount][SpvProfileCount][EShLangCount] = {};
279c292,293
<     if (CommonSymbolTable[versionIndex][profileIndex][EPcGeneral]) {
---
> 	int spvIndex = MapSpvVersionToIndex(spvVersion);
>     if (CommonSymbolTable[versionIndex][profileIndex][spvIndex][EPcGeneral]) {
307,309c321,323
<             CommonSymbolTable[versionIndex][profileIndex][precClass] = new TSymbolTable;
<             CommonSymbolTable[versionIndex][profileIndex][precClass]->copyTable(*commonTable[precClass]);
<             CommonSymbolTable[versionIndex][profileIndex][precClass]->readOnly();
---
>             CommonSymbolTable[versionIndex][profileIndex][spvIndex][precClass] = new TSymbolTable;
>             CommonSymbolTable[versionIndex][profileIndex][spvIndex][precClass]->copyTable(*commonTable[precClass]);
>             CommonSymbolTable[versionIndex][profileIndex][spvIndex][precClass]->readOnly();
314,317c328,331
<             SharedSymbolTables[versionIndex][profileIndex][stage] = new TSymbolTable;
<             SharedSymbolTables[versionIndex][profileIndex][stage]->adoptLevels(*CommonSymbolTable[versionIndex][profileIndex][CommonIndex(profile, (EShLanguage)stage)]);
<             SharedSymbolTables[versionIndex][profileIndex][stage]->copyTable(*stageTables[stage]);
<             SharedSymbolTables[versionIndex][profileIndex][stage]->readOnly();
---
>             SharedSymbolTables[versionIndex][profileIndex][spvIndex][stage] = new TSymbolTable;
>             SharedSymbolTables[versionIndex][profileIndex][spvIndex][stage]->adoptLevels(*CommonSymbolTable[versionIndex][profileIndex][spvIndex][CommonIndex(profile, (EShLanguage)stage)]);
>             SharedSymbolTables[versionIndex][profileIndex][spvIndex][stage]->copyTable(*stageTables[stage]);
>             SharedSymbolTables[versionIndex][profileIndex][spvIndex][stage]->readOnly();
619a634
> 												  [MapSpvVersionToIndex(spvVersion)]
1053,1056c1068,1073
<             for (int lang = 0; lang < EShLangCount; ++lang) {
<                 delete SharedSymbolTables[version][p][lang];
<                 SharedSymbolTables[version][p][lang] = 0;
<             }
---
> 			for (int spvIndex = 0; spvIndex < SpvProfileCount; spvIndex++) {
> 				for (int lang = 0; lang < EShLangCount; ++lang) {
> 	                delete SharedSymbolTables[version][p][spvIndex][lang];
> 		            SharedSymbolTables[version][p][spvIndex][lang] = 0;
> 			    }
> 			}
1062,1065c1079,1084
<             for (int pc = 0; pc < EPcCount; ++pc) {
<                 delete CommonSymbolTable[version][p][pc];
<                 CommonSymbolTable[version][p][pc] = 0;
<             }
---
> 			for (int spvIndex = 0; spvIndex < SpvProfileCount; spvIndex++) {
> 				for (int pc = 0; pc < EPcCount; ++pc) {
> 	                delete CommonSymbolTable[version][p][spvIndex][pc];
> 		            CommonSymbolTable[version][p][spvIndex][pc] = 0;
> 			    }
> 			}

```

**Comment by johnkslang** [comment]
This is along the correct line of what is needed.  I was expected a separate change to solve this, but it hasn't shown up yet, so I will do the above (or similar) soon.

**Comment by johnkslang** [comment]
This should be fixed now by 10119719813e8624531023904304cf1ffb372bd5.  Let me know if you have any other problems on this issue.

**Comment by danginsburg** [comment]
Thank you @johnkslang!
